src/cmdline/mod.rs

```diff
@@ -343,17 +431,106 @@ impl Cmdline {
+    /// # use linux_loader::cmdline::*;
+    /// let cl = Cmdline::try_from("foo -- bar", 100).unwrap();
+    /// assert_eq!(
+    ///     cl.as_cstring().unwrap().as_bytes_with_nul(),
+    ///     b"foo -- bar\0"
+    /// );
+    /// ```
+    pub fn try_from(cmdline_raw: &str, capacity: usize) -> Result<Cmdline> {
+        // The cmdline_raw argument should contain no more than one INIT_ARGS_SEPARATOR sequence
+        // that is not double quoted; in case the INIT_ARGS_SEPARATOR is found all chars following
+        // it will be parsed as init args.
+
+        if capacity == 0 {
+            return Err(Error::InvalidCapacity);
+        }
+
+        // Step 1: Extract boot args and init args from input by searching for INIT_ARGS_SEPARATOR.
+
+        // Check first occurrence of the INIT_ARGS_SEPARATOR that is not between double quotes.
+        // All chars following the INIT_ARGS_SEPARATOR will be parsed as init args.
+        let (mut boot_args, mut init_args) = match cmdline_raw
+            .match_indices(INIT_ARGS_SEPARATOR)
+            .find(|&separator_occurrence| {
+                Self::check_outside_double_quotes(&cmdline_raw[..(separator_occurrence.0)])
+            }) {
+            None => (cmdline_raw, ""),
+            Some((delimiter_index, _)) => (
+                &cmdline_raw[..delimiter_index],
+                // This does not overflow as long as `delimiter_index + INIT_ARGS_SEPARATOR.len()`
+                // is pointing to the first char after the INIT_ARGS_SEPARATOR which always exists;
+                // as a result, `delimiter_index + INIT_ARGS_SEPARATOR.len()` is less or equal to the
+                // length of the initial string.
+                &cmdline_raw[(delimiter_index + INIT_ARGS_SEPARATOR.len())..],
+            ),
+        };
+
+        boot_args = boot_args.trim();
+        init_args = init_args.trim();
+
+        // Step 2: Check if capacity provided for the cmdline is not exceeded and create a new `Cmdline`
+        // if size check passes.
+        let mut cmdline_size = boot_args.len().checked_add(1).ok_or(Error::TooLarge)?;
+
+        if !init_args.is_empty() {
+            cmdline_size = cmdline_size
+                .checked_add(INIT_ARGS_SEPARATOR.len())
+                .ok_or(Error::TooLarge)?;
+
+            cmdline_size = cmdline_size
+                .checked_add(init_args.len())
+                .ok_or(Error::TooLarge)?;
+        }
+
+        if cmdline_size <= capacity {
+            Ok(Cmdline {
+                boot_args: boot_args.to_string(),
+                init_args: init_args.to_string(),
+                capacity,
+            })
+        } else {
+            Err(Error::InvalidCapacity)
+        }
+    }
 }
 
 impl From<Cmdline> for Vec<u8> {
     fn from(cmdline: Cmdline) -> Vec<u8> {
-        cmdline.line.into_bytes()
+        match cmdline.as_cstring() {
+            Ok(cmdline_cstring) => cmdline_cstring.into_bytes_with_nul(),
+            Err(_) => Vec::from([0u8]),
```

Comment:
This is a silent failure. We should change this to try_from instead and return the error from `into_bytes_with_nul` instead.

Reply:
solved